File: DotNETEntity/DataLoader.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Data.SQLite;


namespace DatabaseDemo {

public delegate T RowCallback<T>(SQLiteDataReader reader);

public class DataLoader : IDisposable {
    string DbPath { get; init; }
    SQLiteConnection Conn { get; init; }
    bool IsDisposed { get; set; }

    public DataLoader(string dbPath) {
        DbPath = dbPath;
        Conn = new SQLiteConnection($"Data Source={dbPath}");
        Conn.Open();
        IsDisposed = false;
    }

    public List<T> LoadAll<T>(string sql, RowCallback<T> callback) {
        var cmd = new SQLiteCommand(sql, Conn);
        var reader = cmd.ExecuteReader();
        var rows = new List<T>();

        while (reader.Read()) {
            rows.Add(callback(reader));
        }

        return rows;
    }

    public List<Region> LoadRegions() {
        // Use recursive CTE to query the hierarchy in the correct order,
        // starting with the parents and descending down level-by-level.
        // This is needed for upholding the foreign key constraints when
        // re-inserting the data into the schema of EntityFramework Core.
        var regions = new Dictionary<long, Region>();
        var sql = @"
            WITH tmp AS (
                    SELECT id, name, parent_id, CAST(id AS TEXT) AS path
                    FROM region
                    WHERE parent_id IS NULL
                UNION ALL
                    SELECT region.id, region.name, region.parent_id,
                           (tmp.path || '/' || CAST(region.id AS TEXT)) AS path
                    FROM region
                    INNER JOIN tmp
                    ON region.parent_id = tmp.id
            )
            SELECT id, name, parent_id
            FROM tmp
            ORDER BY path
        ";

        var regionList = LoadAll(sql, reader => {
            var parentId = reader.IsDBNull(2) ? null : reader.GetInt64(2) as long?;
            var parent = reader.IsDBNull(2) ? null : regions[reader.GetInt64(2)];
            var region = new Region {
                RegionId = reader.GetInt64(0),
                Name = reader.GetString(1),
                Bounds = new List<Location>(),
                ParentId = parentId,
                Parent = parent,
            };
            regions[region.RegionId] = region;
            return region;
        });

        // populate regions with locations
        LoadLocations(regionList);

        return regionList;
    }

    public List<Location> LoadLocations(List<Region> regions) {
        var regionsById = regions.ToDictionary(region => region.RegionId, region => region);
        var sql = "SELECT id, latitude, longitude, region_id FROM location ORDER BY region_id, id";
        return LoadAll(sql, reader => {
            var region = regionsById[reader.GetInt64(3)];
            var location = new Location {
                LocationId = reader.GetInt64(0),
                Latitude = reader.GetDouble(1),
                Longitude = reader.GetDouble(2),
                Region = region,
            };
            region.Bounds.Add(location);
            return location;
        });
    }

    public List<Session> LoadSessions() {
        var sql = @"
            SELECT id, user_id,
                   login_public_key, login_date,
                   logout_public_key, logout_date
            FROM session
            ORDER BY user_id, id
        ";

        return LoadAll(sql, reader => {
            var logoutEvent = reader.IsDBNull(4) ? null : new AuthEvent {
                PublicKey = reader.GetByteArray(4),
                Date = reader.GetDateTime(5),
            };
            return new Session {
                SessionId = reader.GetInt64(0),
                UserId = reader.GetInt64(1),
                LoginEvent = new AuthEvent {
                    PublicKey = reader.GetByteArray(2),
                    Date = reader.GetDateTime(3),
                },
                LogoutEvent = logoutEvent,
            };
        });
    }

    public List<Profile> LoadProfiles() {
        var profiles = new List<Profile>();

        foreach (var p in LoadFacebookProfiles()) {
            profiles.Add(p);
        }
        foreach (var p in LoadGoogleProfiles()) {
            profiles.Add(p);
        }
        foreach (var p in LoadInternalProfiles()) {
            profiles.Add(p);
        }

        return profiles;
    }

    public List<FacebookProfile> LoadFacebookProfiles() {
        var sql = @"
            SELECT profile.id, profile.user_id, profile_facebook.facebook_account_id
            FROM profile
            INNER JOIN profile_facebook
            ON profile.facebook_id = profile_facebook.id
            ORDER BY profile.user_id, profile.id
        ";

        return LoadAll(sql, reader => {
            return new FacebookProfile {
                ProfileId = reader.GetInt64(0),
                UserId = reader.GetInt64(1),
                FacebookAccountId = reader.GetString(2),
            };
        });
    }

    public List<GoogleProfile> LoadGoogleProfiles() {
        var sql = @"
            SELECT profile.id,
                   profile.user_id,
                   profile_google.google_account_id,
                   profile_google.email,
                   profile_google.image_url
            FROM profile
            INNER JOIN profile_google
            ON profile.google_id = profile_google.id
            ORDER BY profile.user_id, profile.id
        ";

        return LoadAll(sql, reader => {
            return new GoogleProfile {
                ProfileId = reader.GetInt64(0),
                UserId = reader.GetInt64(1),
                GoogleAccountId = reader.GetString(2),
                EmailAddress = reader.IsDBNull(3) ? null : reader.GetString(3),
                ImageUrl = reader.IsDBNull(4) ? null : reader.GetString(4),
            };
        });
    }

    public List<InternalProfile> LoadInternalProfiles() {
        var sql = @"
            SELECT profile.id,
                   profile.user_id,
                   profile_internal.password_hash,
                   profile_internal.password_salt
            FROM profile
            INNER JOIN profile_internal
            ON profile.internal_id = profile_internal.id
            ORDER BY profile.user_id, profile.id
        ";

        return LoadAll(sql, reader => {
            return new InternalProfile {
                ProfileId = reader.GetInt64(0),
                UserId = reader.GetInt64(1),
                PasswordHash = reader.GetByteArray(2),
                PasswordSalt = reader.GetByteArray(3),
            };
        });
    }

    public List<User> LoadUsers(List<Profile> profiles, List<Session> sessions) {
        var profilesByUser = profiles
            .GroupBy(x => x.UserId)
            .ToDictionary(x => x.Key, x => x.ToList());
        var sessionsByUser = sessions
            .GroupBy(x => x.UserId)
            .ToDictionary(x => x.Key, x => x.ToList());

        var sql = @"
            SELECT id, username, real_name, birth_date
            FROM user
            ORDER BY id
        ";

        // In the two lines below marked "deferred", we do not
        // yet populate the owned real estates and sessions of
        // the user, because the "main" direction of these
        // connections is RealEstate -> Owner and Booking -> User.
        // These two properties are only created for the sake of
        // efficiency, because EF Core generates better queries
        // if we use the list navigation properties than it would
        // if we wrote explicit subqueries.
        return LoadAll(sql, reader => {
            var userId = reader.GetInt64(0);
            return new User {
                UserId = userId,
                Username = reader.GetString(1),
                RealName = reader.IsDBNull(2) ? null : reader.GetString(2),
                BirthDate = reader.IsDBNull(3) ? null : reader.GetDateTime(3),
                Profiles = profilesByUser.GetOrDefault(userId, new List<Profile>()),
                Sessions = sessionsByUser.GetOrDefault(userId, new List<Session>()),
                OwnedRealEstates = new List<RealEstate>(), // deferred
                Bookings = new List<Booking>(), // deferred
            };
        });
    }

    public List<RealEstate> LoadRealEstates(List<User> users, List<Region> regions) {
        var usersById = users.ToDictionary(u => u.UserId, u => u);
        var regionsById = regions.ToDictionary(r => r.RegionId, r => r);

        return LoadAll("SELECT * FROM real_estate ORDER BY id", reader => {
            return new RealEstate {
                RealEstateId = reader.GetInt64(0),
                Kind = Enum.Parse<RealEstateKind>(reader.GetString(1), true),
                Owner = usersById[reader.GetInt64(2)],
                Region = regionsById[reader.GetInt64(3)],
            };
        });
    }

    public List<Booking> LoadBookings(List<User> users, List<RealEstate> realEstates) {
        var usersById = users.ToDictionary(u => u.UserId, u => u);
        var realEstatesById = realEstates.ToDictionary(r => r.RealEstateId, r => r);

        return LoadAll("SELECT * FROM booking ORDER BY id", reader => {
            return new Booking {
                BookingId = reader.GetInt64(0),
                RealEstate = realEstatesById[reader.GetInt64(1)],
                User = usersById[reader.GetInt64(2)],
                StartDate = reader.GetDateTime(3),
                EndDate = reader.GetDateTime(4),
                Price = reader.GetDecimal(5),
            };
        });
    }

    public List<Duration> LoadDurations() {
        return new List<Duration> {
            new Duration { DurationId = 1, Name = "days"      },
            new Duration { DurationId = 2, Name = "weeks"     },
            new Duration { DurationId = 3, Name = "months"    },
            new Duration { DurationId = 4, Name = "years"     },
            new Duration { DurationId = 5, Name = "centuries" },
        };
    }

    public void PopulateDatabase(DemoDbContext db) {
        db.Database.EnsureCreated();

        // First, load the example data using raw SQLite
        Console.WriteLine("Loading Data");

        var regions = LoadRegions();
        var profiles = LoadProfiles();
        var sessions = LoadSessions();
        var users = LoadUsers(profiles, sessions);
        var realEstates = LoadRealEstates(users, regions);
        var bookings = LoadBookings(users, realEstates);
        var durations = LoadDurations();

        // Then, initialize and populate the DB managed by the ORM
        Console.WriteLine("Inserting Data");

        db.AddRange(regions);
        db.AddRange(users);
        db.AddRange(realEstates);
        db.AddRange(bookings);
        db.AddRange(durations);

        db.SaveChanges();
    }

    public void Dispose() {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing) {
        if (!IsDisposed) {
            if (disposing) {
                Conn.Dispose();
            }

            IsDisposed = true;
        }
    }

    ~DataLoader() {
        Dispose(false);
    }
}

static class SQLiteDataReaderExtension {
    public static byte[] GetByteArray(this SQLiteDataReader reader, int index) {
        var stream = reader.GetStream(index);
        var maybeMemoryStream = stream as MemoryStream;
        if (maybeMemoryStream != null) {
            return maybeMemoryStream.ToArray();
        }

        var memoryStream = new MemoryStream();
        stream.CopyTo(memoryStream);
        return memoryStream.ToArray();
    }
}

}

File: SwiftCoreStore/Sources/DataLoader.swift
import Foundation
import CoreStore

/// Class for retrieving data from the raw SQLite
/// example database and putting it into CoreStore.
final class DataLoader {
    private let sqlDB: SQLiteDB
    private let dataStack: DataStack

    init(sqlDB: SQLiteDB, dataStack: DataStack) {
        self.sqlDB = sqlDB
        self.dataStack = dataStack
    }

    func load() throws {
        try dataStack.perform(synchronous: {
            let users = try loadUsers(transaction: $0)
            let regions = try loadRegions(transaction: $0)
            let realEstates = try loadRealEstates(transaction: $0, users: users, regions: regions)
            try loadBookings(transaction: $0, users: users, realEstates: realEstates)
        })
    }

    private func loadProfiles(transaction: SynchronousDataTransaction) throws -> [Int64: [Profile]] {
        let profileDicts = try sqlDB.query("""
        SELECT profile.*,
               profile_facebook.facebook_account_id,
               profile_google.google_account_id,
               profile_google.email,
               profile_google.image_url,
               profile_internal.password_hash,
               profile_internal.password_salt
        FROM profile
        LEFT JOIN profile_facebook
          ON profile.facebook_id = profile_facebook.id
        LEFT JOIN profile_google
          ON profile.google_id = profile_google.id
        LEFT JOIN profile_internal
          ON profile.internal_id = profile_internal.id
        """)

        let profileDictsByUser = Dictionary(grouping: profileDicts, by: {
            $0["user_id"] as! Int64
        })
        var profilesByUser = [Int64: [Profile]](minimumCapacity: profileDictsByUser.count)

        for (userID, profileDicts) in profileDictsByUser {
            var profiles = [Profile]()
            profiles.reserveCapacity(profileDicts.count)

            for p in profileDicts {
                let profile: Profile

                if p["facebook_id"]! != nil {
                    let tmp = transaction.create(Into<FacebookProfile>())
                    tmp.accountID = p["facebook_account_id"] as! String
                    profile = tmp
                } else if p["google_id"]! != nil {
                    let tmp = transaction.create(Into<GoogleProfile>())
                    tmp.accountID = p["google_account_id"] as! String
                    tmp.email = p["email"] as? String
                    tmp.imageURL = (p["image_url"] as? String).flatMap(URL.init(string:))
                    profile = tmp
                } else if p["internal_id"]! != nil {
                    let tmp = transaction.create(Into<InternalProfile>())
                    tmp.passwordHash = p["password_hash"] as! Data
                    tmp.passwordSalt = p["password_salt"] as! Data
                    profile = tmp
                } else {
                    throw SQLiteError.queryFailed(reason: "invalid Profile variant")
                }

                profiles.append(profile)
            }

            profilesByUser[userID] = profiles
        }

        return profilesByUser
    }

    private func loadSessions(transaction: SynchronousDataTransaction) throws -> [Int64: [Session]] {
        let sessionDicts = try sqlDB.query("""
        SELECT user_id,
               login_public_key, login_date,
               logout_public_key, logout_date
        FROM session
        """)

        let sessionDictsByUser = Dictionary(grouping: sessionDicts, by: {
            $0["user_id"] as! Int64
        })
        var sessionsByUser = [Int64: [Session]](minimumCapacity: sessionDictsByUser.count)

        for (userID, sessionDicts) in sessionDictsByUser {
            var sessions = [Session]()
            sessions.reserveCapacity(sessionDicts.count)

            for s in sessionDicts {
                let login = transaction.create(Into<AuthEvent>())
                login.publicKey = s["login_public_key"] as! Data
                login.date = sqlDB.dateFormatter.date(from: s["login_date"] as! String)!

                let logout: AuthEvent?
                if let logoutPubkey = s["logout_public_key"] as? Data,
                   let logoutDate = s["logout_date"] as? String
                {
                    let tmp = transaction.create(Into<AuthEvent>())
                    tmp.publicKey = logoutPubkey
                    tmp.date = sqlDB.dateFormatter.date(from: logoutDate)!
                    logout = tmp
                } else {
                    logout = nil
                }

                let session = transaction.create(Into<Session>())
                session.login = login
                session.logout = logout
                sessions.append(session)
            }

            sessionsByUser[userID] = sessions
        }

        return sessionsByUser
    }

    private func loadUsers(transaction: SynchronousDataTransaction) throws -> [Int64: User] {
        let profiles = try loadProfiles(transaction: transaction)
        let sessions = try loadSessions(transaction: transaction)

        let userDicts = try sqlDB.query("""
        SELECT id, username, real_name, birth_date
        FROM user
        """)
        var usersByID = [Int64: User](minimumCapacity: userDicts.count)

        for u in userDicts {
            let user = transaction.create(Into<User>())

            user.userID = u["id"] as! Int64
            user.username = u["username"] as! String
            user.realName = u["real_name"] as? String
            user.birthDate = (u["birth_date"] as? String).flatMap(sqlDB.dateFormatter.date(from:))
            user.profiles = profiles[user.userID] ?? []
            user.sessions = sessions[user.userID] ?? []

            usersByID[user.userID] = user
        }

        return usersByID
    }

    private func loadLocations(transaction: SynchronousDataTransaction) throws -> [Int64: [Location]] {
        let locationDicts = try sqlDB.query(
            "SELECT latitude, longitude, region_id FROM location ORDER BY region_id"
        )
        let locationDictsByRegion = Dictionary(grouping: locationDicts, by: {
            $0["region_id"] as! Int64
        })
        var locationsByRegion = [Int64: [Location]](minimumCapacity: locationDictsByRegion.count)

        for (regionID, locationDicts) in locationDictsByRegion {
            var locations = [Location]()
            locations.reserveCapacity(locationDicts.count)

            for loc in locationDicts {
                let location = transaction.create(Into<Location>())
                location.latitude = loc["latitude"] as! Double
                location.longitude = loc["longitude"] as! Double
                locations.append(location)
            }

            locationsByRegion[regionID] = locations
        }

        return locationsByRegion
    }

    private func loadRegions(transaction: SynchronousDataTransaction) throws -> [Int64: Region] {
        // Select regions one level at a time, starting with the parents.
        // In this way, we can keep updating our region ID -> Region object
        // mapping, `regionsByID`, and always ensure we have the parents
        // in the dictionary by the time their children need them.
        let regionDicts = try sqlDB.query("""
        WITH tmp AS (
                SELECT id, name, parent_id, CAST(id AS TEXT) AS path
                FROM region
                WHERE parent_id IS NULL
            UNION ALL
                SELECT region.id, region.name, region.parent_id,
                       (tmp.path || '/' || CAST(region.id AS TEXT)) AS path
                FROM region
                INNER JOIN tmp
                ON region.parent_id = tmp.id
        )
        SELECT id, name, parent_id
        FROM tmp
        ORDER BY path
        """)
        var regionsByID = [Int64: Region](minimumCapacity: regionDicts.count)
        let locationsByRegion = try loadLocations(transaction: transaction)

        for r in regionDicts {
            let region = transaction.create(Into<Region>())
            region.name = r["name"] as! String
            let regionID = r["id"] as! Int64
            let parentID = r["parent_id"] as? Int64
            region.regionID = regionID
            region.bounds = locationsByRegion[regionID] ?? []
            region.parent = parentID.flatMap { regionsByID[$0] }
            regionsByID[regionID] = region
        }

        return regionsByID
    }

    private func loadRealEstates(
        transaction: SynchronousDataTransaction,
        users: [Int64: User],
        regions: [Int64: Region]
    ) throws -> [Int64: RealEstate] {
        let realEstateDicts = try sqlDB.query("SELECT * FROM real_estate")
        var realEstatesByID = [Int64: RealEstate](minimumCapacity: realEstateDicts.count)

        for r in realEstateDicts {
            let realEstateID = r["id"] as! Int64
            let ownerID = r["owner_id"] as! Int64
            let regionID = r["region_id"] as! Int64
            let realEstate = transaction.create(Into<RealEstate>())
            realEstate.kind = RealEstateKind(rawValue: r["kind"] as! String)!
            realEstate.owner = users[ownerID]!
            realEstate.region = regions[regionID]!
            realEstatesByID[realEstateID] = realEstate
        }

        return realEstatesByID
    }

    private func loadBookings(
        transaction: SynchronousDataTransaction,
        users: [Int64: User],
        realEstates: [Int64: RealEstate]
    ) throws {
        let bookingDicts = try sqlDB.query("SELECT * FROM booking")

        for b in bookingDicts {
            let userID = b["user_id"] as! Int64
            let realEstateID = b["real_estate_id"] as! Int64
            let startDate = b["start_date"] as! String
            let endDate = b["end_date"] as! String

            let price: NSDecimalNumber
            if let p = b["price"] as? Int64 {
                price = NSDecimalNumber(value: p)
            } else {
                price = NSDecimalNumber(value: b["price"] as! Double)
            }

            let booking = transaction.create(Into<Booking>())
            booking.user = users[userID]!
            booking.realEstate = realEstates[realEstateID]!
            booking.startDate = sqlDB.dateFormatter.date(from: startDate)!
            booking.endDate = sqlDB.dateFormatter.date(from: endDate)!
            booking.category = BookingLengthCategory(startDate: booking.startDate, endDate: booking.endDate)
            booking.price = price
        }
    }
}
